Nama: Joyce Paulina Simanjuntak NIM: 42324059 Prodi: D3 Teknologi Informasi

# Naïve Bayes  

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

Cell ini digunakan untuk mengimpor library yang dibutuhkan, yaitu NumPy, Pandas, dan load_iris dari Scikit-learn. NumPy dan Pandas digunakan untuk membantu pengolahan data, sedangkan load_iris berfungsi untuk memuat Dataset Iris yang berisi karakteristik tiga jenis bunga, yaitu Setosa, Versicolor, dan Virginica. Dataset ini nantinya akan digunakan sebagai data untuk proses klasifikasi.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Cell ini digunakan untuk mengubah Dataset Iris ke dalam bentuk tabel menggunakan Pandas agar data lebih mudah dibaca dan diolah. Pada tabel tersebut, terdapat empat fitur, yaitu Sepal Length, Sepal Width, Petal Length, dan Petal Width, serta jenis bunga yang disimpan dalam kolom Species. Selanjutnya, data dipisahkan menjadi X yang berisi karakteristik bunga sebagai fitur dan y yang berisi jenis bunga sebagai label atau target yang akan diprediksi. Dengan pemisahan ini, data sudah siap digunakan untuk proses klasifikasi.

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Variabel y digunakan untuk menyimpan label atau target dari Dataset Iris, yaitu jenis bunga yang terdiri dari Setosa, Versicolor, dan Virginica. Data ini nantinya menjadi acuan bagi model dalam mempelajari dan memprediksi jenis bunga berdasarkan karakteristik yang terdapat pada variabel X.

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Bagian ini digunakan untuk mengubah label pada y yang sebelumnya berupa nama bunga menjadi angka agar lebih mudah diproses oleh model machine learning. LabelEncoder akan memberikan kode angka untuk setiap jenis bunga, sehingga Setosa, Versicolor, dan Virginica berubah menjadi label numerik. Hasil akhirnya disimpan kembali ke variabel y sebagai target yang siap digunakan dalam proses klasifikasi.

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

Cell ini digunakan untuk membagi data menjadi data training dan data testing. Sebanyak 80% data digunakan sebagai data training untuk membantu model belajar, sedangkan 20% sisanya digunakan sebagai data testing untuk melihat bagaimana model bekerja pada data yang belum digunakan saat proses pembelajaran. random_state=0 digunakan agar pembagian data tetap sama setiap kali kode dijalankan.

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Cell ini digunakan untuk meminta model NB_model melakukan prediksi terhadap data pada X_test. Hasil prediksi tersebut disimpan ke dalam variabel result, yang berisi jenis bunga yang diperkirakan oleh model berdasarkan data pengujian. Dengan begitu, hasil pada result nantinya dapat dibandingkan dengan label sebenarnya pada y_test untuk melihat seberapa baik model melakukan klasifikasi.

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Bagian ini dipakai untuk mengecek hasil kerja model setelah melakukan prediksi. Hasil prediksi result dibandingkan dengan jawaban sebenarnya pada y_test, lalu dihitung berapa banyak yang sesuai. Nilainya ditampilkan dalam bentuk persen supaya lebih gampang melihat tingkat keberhasilan model dalam mengenali jenis bunga Iris.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\User\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Bagian ini untuk mencoba memasukkan data bunga baru dengan nilai panjang dan lebar sepal serta petal yang sudah ditentukan. Data tersebut kemudian diberikan ke NB_model untuk melihat bunga tersebut termasuk jenis apa. Karena hasil dari model masih berupa angka, inverse_transform digunakan untuk mengembalikannya menjadi nama bunga seperti Setosa, Versicolor, atau Virginica. Hasil akhirnya ditampilkan sebagai jenis bunga yang diprediksi oleh model.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Data baru yang dimasukkan memiliki ukuran sepal 5.1 × 3.5 dan petal 1.4 × 0.2. Data tersebut kemudian dibuat dalam bentuk DataFrame agar sesuai dengan format data Iris sebelumnya. Selanjutnya, NB_model digunakan untuk menebak jenis bunga dari data tersebut. Karena hasil prediksi masih berupa angka, inverse_transform mengubahnya kembali menjadi nama spesies sehingga hasilnya lebih mudah dibaca.

# TUGAS

In [24]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-learn
iris = load_iris()

Sel ini seperti menyiapkan perlengkapan sebelum masak. numpy buat urusan hitung-hitungan angka, pandas buat bikin tabel, lalu kita ambil dataset Iris yang sudah disediakan scikit-learn. Isinya data 150 bunga iris dari 3 jenis (setosa, versicolor, virginica), lengkap dengan ukuran kelopak dan mahkotanya.

In [26]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Data mentahnya di ubah jadi tabel biar enak dilihat, mirip Excel, lalu kita tambah kolom Species yang isinya nama bunga. Setelah itu tabelnya kita pisah dua. X adalah "soal" (4 ukuran bunga: panjang dan lebar kelopak, panjang dan lebar mahkota), dan y adalah "kunci jawaban" (nama spesiesnya). Hasil print(X) menunjukkan 150 baris dan 4 kolom.

In [27]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Cuma buat ngecek isi y. Di sini masih berupa tulisan: setosa, versicolor, virginica.

In [28]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Komputer lebih nyaman kerja dengan angka daripada tulisan. Jadi LabelEncoder mengubah setosa jadi 0, versicolor jadi 1, dan virginica jadi 2. Nanti kalau hasil prediksinya sudah keluar, kita tinggal balikin lagi ke nama aslinya.

In [29]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Di bagi 150 data jadi dua kelompok. 80% (120 data) buat latihan model, dan 20% (30 data) buat ujian. Kenapa dipisah? Biar adil, karena model diuji pakai soal yang belum pernah dia lihat. random_state=0 bikin pembagian acaknya selalu sama, jadi hasil kita akan sama dengan teman-teman.

In [30]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
NB_model = model.fit(X_train, y_train)

Ini bagian intinya. Di sini pakai Gaussian Naïve Bayes karena ukuran bunga itu angka kontinu (ada koma-komanya). Lewat fit(), model mempelajari 120 data latih. Dia mencatat rata-rata dan sebaran ukuran bunga untuk tiap spesies, misalnya setosa biasanya mungil dan virginica biasanya besar.

In [31]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Sekarang 30 data ujian kita kasih ke model, dan dia menebak spesiesnya. Hasilnya berupa deretan angka 0, 1, 2 yang mewakili spesies tadi.

In [32]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Tebakan model dicocokkan dengan kunci jawaban asli. Hasilnya 96,67%, artinya dari 30 soal, 29 dijawab benar dan cuma 1 yang meleset. Nilai yang bagus.

In [33]:
#prediksi inputan data baru
input_data = [[7.3, 3.8, 4.6, 2.3]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


c:\Users\User\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Ada bunga baru dengan ukuran 7.3 , 3.8 , 4.6 , 2.3, dan model diminta menebak jenisnya. Hasil tebakannya masih angka, jadi inverse_transform mengubahnya balik jadi nama spesies. Nanti akan muncul tulisan merah UserWarning. Ini bukan error, itu cuma pengingat karena model dilatih dengan data yang punya nama kolom, sedangkan inputan nya berupa list biasa tanpa nama kolom.

In [34]:
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Ukuran bunga yang dimasukkan adalah 7.3, 3.8, 4.6, dan 2.3 untuk masing-masing fitur sepal dan petal. Nilai tersebut kemudian diproses oleh NB_model untuk menentukan jenis bunga berdasarkan pola yang sudah dipelajari sebelumnya. Hasil prediksi yang berupa angka diubah kembali menjadi nama spesies menggunakan inverse_transform, kemudian ditampilkan sebagai hasil prediksi.

# Kesimpulan
Gaussian Naïve Bayes dipakai karena semua fitur di dataset Iris berupa angka kontinu. Model dilatih dengan 120 data dan diuji dengan 30 data, dengan akurasi 96,67%, jadi cukup andal untuk mengklasifikasikan bunga iris.

Untuk data baru [7.3, 3.8, 4.6, 2.3], model memprediksi bunga tersebut termasuk species virginica. Hasil ini masuk akal karena ukuran mahkotanya besar (petal width 2,3 cm), berbeda intinya dengan setosa yang bunganya kecil.